packages/react-admin/src/AdminRoutes.js

```diff
@@ -40,98 +33,115 @@ export class AdminRoutes extends Component {
         }
     };
 
-    async initializeResources() {
-        const { children, declareResources } = this.props;
-        const { permissions } = this.state;
-
-        if (typeof children === 'function') {
-            const childrenResult = children(permissions);
-            let resources = childrenResult;
-
-            if (typeof childrenResult.then === 'function') {
-                resources = await childrenResult;
+    componentWillUpdate(nextProps, nextState) {
+        if (
+            typeof nextProps.children === 'function' &&
+            nextState.permissions !== this.state.permissions
+        ) {
+            const { children } = nextProps;
+
+            const childrenFuncResult = children(nextState.permissions);
+            if (childrenFuncResult.then) {
+                childrenFuncResult.then(resolvedChildren => {
+                    this.setState({
```

Comment:
Is it a good practice to update the state (`childrenToRender`) while it was just updated (`premissions`)? Won't it cause a double render?

Reply:
Yes but it only renders a Loading screen if there are no children to render anyway. Are you worried about updated permissions ? In this case, yes too but I don't see any other way to do it